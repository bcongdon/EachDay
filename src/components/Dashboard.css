.react-calendar-heatmap .color-scale-1 { fill: #e50700; }
.react-calendar-heatmap .color-scale-2 { fill: #e04500; }
.react-calendar-heatmap .color-scale-3 { fill: #dc8000; }
.react-calendar-heatmap .color-scale-4 { fill: #d8b900; }
.react-calendar-heatmap .color-scale-5 { fill: #b8d400; }
.react-calendar-heatmap .color-scale-6 { fill: #7bcf00; }
.react-calendar-heatmap .color-scale-7 { fill: #41cb00; }
.react-calendar-heatmap .color-scale-8 { fill: #09c700; }
.react-calendar-heatmap .color-scale-9 { fill: #00c32c; }
.react-calendar-heatmap .color-scale-10 { fill: #00bf60; }