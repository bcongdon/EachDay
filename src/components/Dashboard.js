import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import CalendarHeatmap from 'react-calendar-heatmap'
import { chain } from 'lodash'
import './Dashboard.css'
import { PropTypes } from 'prop-types'
import { loadEntries, openEntryModal } from '../actions'
import UserNavbar from './UserNavbar'
import EntryModal from './entry/EntryModal'
import Entry from './entry/Entry'
import { Button, Grid, Divider, Message } from 'semantic-ui-react'

class Dashboard extends Component {
  componentWillMount() {
    this.props.loadEntries()
  }

  getEntries() {
    const entries = this.props.entries.map((e) => {
      return (<Entry key={e.id} rating={e.rating} date={e.date} notes={e.notes} />)
    })

    if (entries.length === 0) {
      return <Message compact>You don't have any entries yet!</Message>
    } else {
      return entries
    }
  }

  render() {
    const values = chain(this.props.entries)
    .map(e => {
      return {
        date: e.date,
        count: e.rating,
        title: e.notes
      }
    })
    .filter(e => { return e.count !== undefined })
    .value()

    const composeEntryButton = (
      <Button
        icon='plus'
        content='Compose Entry'
        onClick={() => this.props.openEntryModal()}
        />
    )

    return (
      <div>
        <UserNavbar />
        <Grid centered verticalAlign='middle' columns={1}>
          <Grid.Column style={{'maxWidth': 600}}>
            <EntryModal trigger={composeEntryButton} />
            <Divider />
            <CalendarHeatmap
              numDays={366}
              values={values}
              classForValue={(value) => {
                if (!value) {
                  return 'color-empty'
                }
                return `color-scale-${value.count}`
              }} />
            <Divider />
            {this.getEntries()}
          </Grid.Column>
        </Grid>
      </div>
    )
  }
}

function mapStateToProps(state) {
  return {
    entries: state.entry.entries
  }
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ loadEntries, openEntryModal }, dispatch)
}

Dashboard.propTypes = {
  loadEntries: PropTypes.func.isRequired,
  openEntryModal: PropTypes.func.isRequired,
  entries: PropTypes.array
}

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard)
