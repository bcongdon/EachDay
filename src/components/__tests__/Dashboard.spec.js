import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Dashboard } from '../Dashboard.js'
import { scroller } from 'react-scroll'
import { Message } from 'semantic-ui-react'

jest.mock('universal-cookie')
jest.mock('react-scroll')

describe('Dashboard', () => {
  beforeEach(() => {
    jest.resetAllMocks()
    Date.now = jest.fn(() => 1495984743025)
  })

  it('should load and show loaders entries on mount', () => {
    const loadEntries = jest.fn()
    var dash = shallow(
      <Dashboard
        loading
        entries={[]}
        openEntryModal={jest.fn()}
        loadEntries={loadEntries}
        />
    )
    expect(loadEntries).toBeCalled()
    expect(dash.find('Loader').length).toEqual(2)
  })

  it('should show message when no entries exist', () => {
    var dash = shallow(
      <Dashboard
        loading={false}
        entries={[]}
        openEntryModal={jest.fn()}
        loadEntries={jest.fn()}
        />
    )
    expect(toJson(dash)).toMatchSnapshot()
    const message = dash.find('Message').render().text()
    expect(message).toEqual(expect.stringContaining('You don\'t have any entries'))
  })

  it('should render errors', () => {
    var dash = shallow(
      <Dashboard
        loading={false}
        entries={[]}
        openEntryModal={jest.fn()}
        loadEntries={jest.fn()}
        error='foo'
        />
    )
    expect(toJson(dash)).toMatchSnapshot()
    const errors = dash.find('ErrorMessage')
    expect(errors.length).toBe(1)
    expect(errors.render().text()).toEqual(expect.stringContaining('foo'))
  })

  it('should have composeEntryButton', () => {
    const openEntryModal = jest.fn()
    var dash = shallow(
      <Dashboard
        entries={[]}
        openEntryModal={openEntryModal}
        loadEntries={jest.fn()}
        />
    )
    const btn = dash.find('Connect(EntryModal)').prop('trigger')
    btn.props.onClick() // 'Click' the trigger button
    expect(openEntryModal).toBeCalled()
  })

  it('should render entries on calendar heatmap', () => {
    const entries = [
      { id: 1, rating: 3, date: '2017-05-01' },
      { id: 2, rating: 5, date: '2017-05-02' },
      { id: 3, rating: 4, date: '2017-05-03' }
    ]
    var dash = shallow(
      <Dashboard
        entries={entries}
        openEntryModal={jest.fn()}
        loadEntries={jest.fn()}
        />
    )
    expect(dash.find('Connect(Entry)').length).toBe(entries.length)
    expect(toJson(dash)).toMatchSnapshot()
  })

  it('should give classes to calendar days correctly', () => {
    var dash = shallow(
      <Dashboard
        entries={[]}
        openEntryModal={jest.fn()}
        loadEntries={jest.fn()}
        />
    )
    const classForValue = dash.instance().classForValue
    expect(classForValue(null)).toBe('ui color-empty')
    expect(classForValue({date: '2017-01-01'})).toBe('ui color-empty')
    expect(classForValue({date: '2017-01-01', count: 5})).toBe('ui color-scale-5')
  })

  it('should scroll correctly when an entry is clicked', () => {
    var dash = shallow(
      <Dashboard
        entries={[]}
        openEntryModal={jest.fn()}
        loadEntries={jest.fn()}
        />
    )
    const onEntryClick = dash.instance().onEntryClick
    onEntryClick({})
    expect(scroller.scrollTo).not.toBeCalled()
    onEntryClick({id: 5})
    expect(scroller.scrollTo).toBeCalled()
    expect(scroller.scrollTo.mock.calls[0]).toEqual(['entry-5', { duration: 250, smooth: true }])
  })

  it('should render custom tooltip titles correctly', () => {
    var dash = shallow(
      <Dashboard
        entries={[]}
        openEntryModal={jest.fn()}
        loadEntries={jest.fn()}
        />
    )
    const customTootipTitle = dash.instance().customTootipTitle
    expect(customTootipTitle({})).toEqual({'data-tip': 'No Entries'})
    expect(customTootipTitle({date: '2017-01-01', count: 5})).toEqual({'data-tip': 'Jan 01: ⭐ ⭐ ⭐ ⭐ ⭐ '})
    expect(customTootipTitle({date: '2017-05-30', count: 1})).toEqual({'data-tip': 'May 30: ⭐ '})
  })

  it('should have link to open entry modal in empty entries message', () => {
    const openEntryModal = jest.fn()
    var dash = shallow(
      <Dashboard
        entries={[]}
        openEntryModal={openEntryModal}
        loadEntries={jest.fn()}
        />
    )
    dash.find(Message).find('a').simulate('click')
    expect(openEntryModal).toBeCalled()
  })
})
