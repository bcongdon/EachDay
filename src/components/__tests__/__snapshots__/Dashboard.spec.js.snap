// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should render entries on calendar heatmap 1`] = `
<Grid
  centered={true}
  verticalAlign="middle"
>
  <GridColumn
    style={
      Object {
        "maxWidth": 950,
      }
    }
  >
    <Connect(EntryModal)
      trigger={
        <Button
          as="button"
          content="Compose Entry"
          icon="plus"
          onClick={[Function]}
        />
      }
    />
    <Divider />
    <DimmerDimmable>
      <Dimmer
        inverted={true}
      >
        <Loader>
          Loading
        </Loader>
      </Dimmer>
      <MediaQuery
        component="div"
        minWidth={720}
        style={
          Object {
            "margin": "0 auto",
            "width": 700,
          }
        }
        values={Object {}}
      >
        <ReactTooltip
          effect="solid"
          insecure={true}
          offset={
            Object {
              "right": 6,
            }
          }
          resizeHide={true}
          wrapper="div"
        />
        <e
          classForValue={[Function]}
          endDate={1495984743025}
          gutterSize={1}
          horizontal={true}
          numDays={365}
          onClick={[Function]}
          showMonthLabels={true}
          showOutOfRangeDays={false}
          tooltipDataAttrs={[Function]}
          values={
            Array [
              Object {
                "count": 3,
                "date": "2017-05-01",
                "id": 1,
              },
              Object {
                "count": 5,
                "date": "2017-05-02",
                "id": 2,
              },
              Object {
                "count": 4,
                "date": "2017-05-03",
                "id": 3,
              },
            ]
          }
        />
        <Divider />
      </MediaQuery>
    </DimmerDimmable>
    <Grid
      divided="vertically"
      style={
        Object {
          "margin": 10,
        }
      }
    >
      <Connect(Entry)
        date="2017-05-01"
        id={1}
        rating={3}
      />
      <Connect(Entry)
        date="2017-05-02"
        id={2}
        rating={5}
      />
      <Connect(Entry)
        date="2017-05-03"
        id={3}
        rating={4}
      />
    </Grid>
  </GridColumn>
</Grid>
`;

exports[`Dashboard should render errors 1`] = `
<Grid
  centered={true}
  verticalAlign="middle"
>
  <GridColumn
    style={
      Object {
        "maxWidth": 950,
      }
    }
  >
    <Connect(EntryModal)
      trigger={
        <Button
          as="button"
          content="Compose Entry"
          icon="plus"
          onClick={[Function]}
        />
      }
    />
    <Divider />
    <DimmerDimmable
      dimmed={false}
    >
      <Dimmer
        active={false}
        inverted={true}
      >
        <Loader>
          Loading
        </Loader>
      </Dimmer>
      <MediaQuery
        component="div"
        minWidth={720}
        style={
          Object {
            "margin": "0 auto",
            "width": 700,
          }
        }
        values={Object {}}
      >
        <ReactTooltip
          effect="solid"
          insecure={true}
          offset={
            Object {
              "right": 6,
            }
          }
          resizeHide={true}
          wrapper="div"
        />
        <e
          classForValue={[Function]}
          endDate={1495984743025}
          gutterSize={1}
          horizontal={true}
          numDays={365}
          onClick={[Function]}
          showMonthLabels={true}
          showOutOfRangeDays={false}
          tooltipDataAttrs={[Function]}
          values={Array []}
        />
        <Divider />
      </MediaQuery>
    </DimmerDimmable>
    <ErrorMessage
      compact={true}
      message="foo"
    />
  </GridColumn>
</Grid>
`;

exports[`Dashboard should show message when no entries exist 1`] = `
<Grid
  centered={true}
  verticalAlign="middle"
>
  <GridColumn
    style={
      Object {
        "maxWidth": 950,
      }
    }
  >
    <Connect(EntryModal)
      trigger={
        <Button
          as="button"
          content="Compose Entry"
          icon="plus"
          onClick={[Function]}
        />
      }
    />
    <Divider />
    <DimmerDimmable
      dimmed={false}
    >
      <Dimmer
        active={false}
        inverted={true}
      >
        <Loader>
          Loading
        </Loader>
      </Dimmer>
      <MediaQuery
        component="div"
        minWidth={720}
        style={
          Object {
            "margin": "0 auto",
            "width": 700,
          }
        }
        values={Object {}}
      >
        <ReactTooltip
          effect="solid"
          insecure={true}
          offset={
            Object {
              "right": 6,
            }
          }
          resizeHide={true}
          wrapper="div"
        />
        <e
          classForValue={[Function]}
          endDate={1495984743025}
          gutterSize={1}
          horizontal={true}
          numDays={365}
          onClick={[Function]}
          showMonthLabels={true}
          showOutOfRangeDays={false}
          tooltipDataAttrs={[Function]}
          values={Array []}
        />
        <Divider />
      </MediaQuery>
    </DimmerDimmable>
    <Message
      compact={true}
    >
      You don't have any entries yet! Why don't you 
      <a
        href="#"
        onClick={[Function]}
      >
        write one
      </a>
      ?
    </Message>
  </GridColumn>
</Grid>
`;
